galley/pkg/runtime/projections/serviceentry/handler.go

```diff
@@ -0,0 +1,311 @@
+	p.lastSnapshotTime = now
+	p.pendingEvents = 0
+
+	b := snapshot.NewInMemoryBuilder()
+
+	// Copy the entries.
+	entries := make([]*mcp.Resource, 0, len(p.mcpResources))
+	for _, r := range p.mcpResources {
+		entries = append(entries, r)
+	}
+
+	// Create the collection resources on the Snapshot.
+	version := fmt.Sprintf("%d", p.version)
+	b.Set(collection.String(), version, entries)
+
+	return b.Build()
+}
+
+func (p *Handler) handleServiceEvent(event resource.Event) {
+	entry := event.Entry
+	fullName := entry.ID.FullName
+
+	switch event.Kind {
+	case resource.Added, resource.Updated:
+
+		// Check to see if the service's version has changed.
+		curEntry, ok := p.services[fullName]
+		if ok && curEntry.ID.Version == entry.ID.Version {
+			scope.Debugf("received service for the current, known version: %v", event)
+			return
+		}
+
+		// Store the service.
+		p.services[fullName] = entry
+
+		// Get the associated endpoints, if available.
+		var endpoints *coreV1.Endpoints
+		if endpointsEntry, ok := p.endpoints[fullName]; ok {
+			endpoints = endpointsEntry.Item.(*coreV1.Endpoints)
+		}
+
+		// Convert to an MCP resource to be used in the snapshot.
+		mcpEntry, ok := p.toMcpResource(entry, endpoints)
+		if !ok {
+			return
+		}
+		p.setMcpEntry(fullName, mcpEntry)
+
+		p.updateVersion()
+
+	case resource.Deleted:
+		// Delete the Service and ServiceEntry
+		delete(p.services, fullName)
+		p.deleteMcpResource(fullName)
+		p.updateVersion()
+	default:
+		scope.Errorf("unknown event kind: %v", event.Kind)
+	}
+}
+
+func (p *Handler) handleEndpointsEvent(event resource.Event) {
+	entry := event.Entry
+	fullName := entry.ID.FullName
+
+	switch event.Kind {
+	case resource.Added, resource.Updated:
+
+		// Check to see if the version has changed.
+		curEntry, ok := p.endpoints[fullName]
+		if ok && curEntry.ID.Version == entry.ID.Version {
+			scope.Debugf("received endpoints for the current, known version: %v", event)
+			return
+		}
+
+		p.endpoints[fullName] = entry
+
+		// Look up the service associated with the endpoints.
+		svcEntry, ok := p.services[fullName]
+		if !ok {
+			// Wait until we get a Service before we create the ServiceEntry.
+			scope.Debugf("received endpoints before service for: %s", fullName)
+			return
+		}
+
+		// Convert to an MCP resource to be used in the snapshot.
+		mcpEntry, ok := p.toMcpResource(svcEntry, entry.Item.(*coreV1.Endpoints))
+		if !ok {
+			return
+		}
+		p.setMcpEntry(fullName, mcpEntry)
+
+		p.updateVersion()
+
+	case resource.Deleted:
+		// The lifecycle of the ServiceEntry is bound to the service, so only delete the endpoints entry here.
+		delete(p.endpoints, fullName)
+
+		// Look up the service associated with the endpoints.
+		svcEntry, ok := p.services[fullName]
+		if !ok {
+			return
+		}
+
+		// Update the MCP entry to clear out the endpoints.
+		mcpEntry, ok := p.toMcpResource(svcEntry, nil)
+		if !ok {
+			return
```

Comment:
I would really like to see a more concrete error log here in addition to the generic one displayed in toMcpResource();

Reply:
what would be more concrete in your mind?

Reply:
Error in delete endpoint with endpoint name, ip , port.